### PR TITLE
feat(zink-codegen): introduce derive macro for events 

### LINKED ISSUE #255: Improve the declaration of events
### Describe the feature

we currently only have a baseline event implementation, gather events as an enum, rich and dynamic `log{1,2,3,4}`

### Solution

```rust
#[derive(Event)]
enum Event {
  Transfer { from: Address, to: Address, value: U256 },
}
```

ABIs are also required

### PR DESCRIPTION
**Pull request by malik672**
Resolves #255 

- [x] generate ABIs for events and gathered events as enums
- [x] introduce util function for converting type to string

### CODE REVIEW

**Comment by clearloop** [comment]
Hi @malik672, could you please follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) in your commit messages, for example for the title

```
# zink is the main package you modified
feat(zink-codegen): introduce derive macro for events 
```

and add details for what you modified in a specific commit message

```
# @4977bd5
-  feat(zink): generate ABIs for events
# @b23c12a
- feat(zink): introduce util function for converting type to string
```

 that would be helpful for reviewing, thank you!

**Comment by malik672** [comment]
> Hi @malik672, could you please follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) in your commit messages, for example for the title
> 
> ```
> # zink is the main package you modified
> feat(zink-codegen): introduce derive macro for events 
> ```
> 
> and add details for what you modified in a specific commit message
> 
> ```
> # @4977bd5
> -  feat(zink): generate ABIs for events
> # @b23c12a
> - feat(zink): introduce util function for converting type to string
> ```
> 
> that would be helpful for reviewing, thank you!

done

**Comment by clearloop** [comment]
> done

Oh sorry, I mean in your commit messages but not the description
Screenshot 2024-11-27 at 13 06 52

I'll help you edit the description, you just need to use the conventional commit messages in your future commits

**Comment by clearloop** [comment]
> apart from the log refactor is anything still remaining for the PR ?

The test is the main thing! because our purpose is introducing a nice developer interface ^ ^

```rust
#[derive(Event)]
pub enum ERC20Events {
   Transfer(from: Address, to: Address, value: U256),
   Approval(owner: Address, spender: Address, value: U256),
   // ...
}

#[zink::external]
fn foo(...) {
   ERC20Events::Transfer(from, to, value).emit()
}
```

If the described interface works, our PR works, we may encounter issues while implementing this interface, and they're the problems we need to solve in this PR, please pay attention to the inline document and [the handbook](https://github.com/zink-lang/zink/blob/main/docs/SUMMARY.md) as well since it will help others using our code!

**Comment by malik672** [comment]
> > apart from the log refactor is anything still remaining for the PR ?
> 
> The test is the main thing! because our purpose is introducing a nice developer interface ^ ^
> 
> ```rust
> #[derive(Event)]
> pub enum ERC20Events {
>    Transfer(from: Address, to: Address, value: U256),
>    Approval(owner: Address, spender: Address, value: U256),
>    // ...
> }
> 
> #[zink::external]
> fn foo(...) {
>    ERC20Events::Transfer(from, to, value).emit()
> }
> ```
> 
> If the described interface works, our PR works, we may encounter issues while implementing this interface, and they're the problems we need to solve in this PR, please pay attention to the inline document and [the handbook](https://github.com/zink-lang/zink/blob/main/docs/SUMMARY.md) as well since it will help others using our code!

ohh ok, will update soon, can i solve the log issue here also ?

**Comment by clearloop** [comment]
> ohh ok, will update soon, can i solve the log issue here also ?

do you have an issue number of the log issue you mentioned? some of the issues may be outdated after our new implementation, feel free to add more `Resolves #xx` in the description of the PR!

on the other hand, we have limited budget issues, it would be proper that not solving two budget issues in one PR ^ ^

**Review by clearloop — COMMENTED** [review]
There will be two steps in this PR:

1. `enum` for static events

```rust
#[zink::Event]
pub enum MyEvent {
  MyStaticEvent1,
  MyStaticEvent2,
}

// which generates

pub struct MyEvent;

impl MyEvent {
  pub fn my_static_event_1() {
     ffi::evm::log0("MyStaticEvent1")
  }

  pub fn my_static_event_2() {
    ffi::evm::log0("My Static Event2")
  }
}
```

2. support dynamic event ( this is the main thing in this PR )

```rust
pub enum MyEvent {
   OneParam(Address), // log1
   TwoParams(Address, U256), // log2
   ThreeParams(Address, Address, U256), // log3
   FourParams(Address, Address, U256, U256) // log4
}

impl MyEvent {
  pub fn one_param(param1: Address) {
     // ... TODO HERE
  }

  pub fn two_params(param1: Address, param2: U256) {
    // ... TODO HERE
  }
  
  // ...
}
```

As we can see, `ffi::evm::log0` only supports static bytes, which means it doesn't work for the dynamic events, so we need to refactor the log interfaces

```rust
//! zink/ffi/evm.rs
extern "C" {
  // ...

   // we keep this log0 because it works perfectly
   fn log0(name: &'static str);
   
   // the trick is here, we use `Bytes32` (each stack item is a bytes32 value in EVM) as the param instead of writing tons of interfaces for support all types
   fn log1(name: &'static str, topic1: Bytes32);
   
   // ...
}
```

so what we need to do now is introducing an interface that maps all types to `Bytes32`, **in this PR, you just need to introduce the interfaces for `Address` and `U256`**, for example for `Address`

https://github.com/zink-lang/zink/blob/e82794761ded3232d029a2d429086bafd5e457ee/zink/src/primitives/address.rs#L35-L44

```rust
impl Asm for Address {
  // ...
  
  #[cfg(target_family = "wasm")]
  fn to_bytes32(&self) -> Bytes32 {
    unsafe { ffi::asm::cast_bytes20(self.0) }
  }
}
```

implementing `cast_bytesN` in https://github.com/zink-lang/zink/blob/main/zink/src/primitives/bytes.rs#L37-L48 to support the logic above.

Update the log exmaple tests, and this is the wrap!

**Comment by malik672** [comment]
@clearloop need some help

File: examples/log.rs [hunk 1]
```diff
@@ -1,99 +1,98 @@
...
-pub fn log0() {
-    Ping.log0();
+pub enum ERC20Events {
+    /// Emitted when tokens are transferred between addresses
+    /// Parameters: from, to, value
+    Transfer(Address, Address, U256),
+
+    /// Emitted when an address approves another address to spend tokens
+    /// Parameters: owner, spender, value
+    Approval(Address, Address, U256),
```
Comment (clearloop) [inline examples/log.rs:20]:
```suggestion
pub enum MyEvent {
    Topic0,
    Topic1(U256),
    Topic2(U256, U256),
    Topic3(U256, U256, U256),
    Topic4(U256, U256, U256, U256)
```

File: examples/log.rs [hunk 2]
```diff
@@ -1,99 +1,98 @@
...
+            event => event.log0(),
+        }
+    }
+
+    /// Log an approval event
+    pub fn log_approval(owner: Address, spender: Address, value: U256) {
+        match Self::Approval(owner, spender, value) {
+            event => event.log0(),
+        }
+    }
```
Comment (clearloop) [inline examples/log.rs:36]:
```suggestion
```

Since we have the proc-macro, don't need to implement them here

File: examples/log.rs [hunk 3]
```diff
@@ -1,99 +1,98 @@
...
+
+        // Test individual event logging
+        erc20_events::log_transfer(from, to, value)?;
+        erc20_events::log_approval(from, spender, value)?;
 
-#[zink::external]
-pub fn log4() {
-    Ping.log4(b"pong", b"ping", b"pong", b"pong");
+        // Test multiple events
+        erc20_events::log_transfer_and_approval(from, to, spender, value)?;
```
Comment (clearloop) [inline examples/log.rs:90]: In the test, you need to call and external call and then in the returned `info`,  check the logs to confirm they are correctly emitted

File: examples/log.rs [hunk 4]
```diff
@@ -1,99 +1,98 @@
...
+        to: Address,
+        spender: Address,
+        value: U256,
+    ) -> Result<(), ()> {
+        // Log transfer first
+        log_transfer(from, to, value);
+        // Then log approval
+        log_approval(from, spender, value);
+        Ok(())
+    }
```
Comment (clearloop) [inline examples/log.rs:70]: The testing interfaces are basically correct! 

for now, let's temporarily forget about the proc-macro, the expanded logic of these methods should be 

```rust
#[zink::external]
pub fn topic0() {
   evm::ffi::log0("MyEvent");
}

#[zink::external]
pub fn topic1(value: U256) {
  ffi::evm::log1("MyEvent", value);
}

// ...
```

I'd suggest to complete this logic first, once it pass, generate them with proc-macro!

**Review by clearloop — COMMENTED** [review]
Basically you're handling too many modules of zink at once )) it's complex tbh

see [my comment on the log example](https://github.com/zink-lang/zink/pull/281#discussion_r1869879264), **let's forget about the proc-macro and go with it first!**


```
RUST_LOG=trace cargo nextest run --release --example log
```
you need to learn about debugging the compiled byteocde with the command above, paste the bytecode to the playground of evm.codes to verify if the generated bytecode are correct

**Comment by malik672** [comment]
@clearloop been looking at this code, still don't know why the macro complains of a U256 to bytes32 mismatch

**Comment by clearloop** [comment]
> @clearloop been looking at this code, still don't know why the macro complains of a U256 to bytes32 mismatch

U256 is not equal to Bytes32, 

```
pub struct U256(Bytes32)
```

mb you local branch forgot to merge the latest main? there was a big update for the bytes types xd

**Comment by malik672** [comment]
> > @clearloop been looking at this code, still don't know why the macro complains of a U256 to bytes32 mismatch
> 
> U256 is not equal to Bytes32,
> 
> ```
> pub struct U256(Bytes32)
> ```
> 
> mb you local branch forgot to merge the latest main? there was a big update for the bytes types xd

ohh maybe that's why but i mean technically it's all the same number of bits

**Comment by clearloop** [comment]
> > > @clearloop been looking at this code, still don't know why the macro complains of a U256 to bytes32 mismatch
> > 
> > 
> > U256 is not equal to Bytes32,
> > ```
> > pub struct U256(Bytes32)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > mb you local branch forgot to merge the latest main? there was a big update for the bytes types xd
> 
> ohh maybe that's why but i mean technically it's all the same number of bits

yes! exactly the same, however we need to introduce interfaces converting them when we need in rust to make rustc happy lmao

**Comment by malik672** [comment]
> > > > @clearloop been looking at this code, still don't know why the macro complains of a U256 to bytes32 mismatch
> > > 
> > > 
> > > U256 is not equal to Bytes32,
> > > ```
> > > pub struct U256(Bytes32)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > mb you local branch forgot to merge the latest main? there was a big update for the bytes types xd
> > 
> > 
> > ohh maybe that's why but i mean technically it's all the same number of bits
> 
> yes! exactly the same, however we need to introduce interfaces converting them when we need in rust to make rustc happy lmao

what do you think remains for this pr, implementation wise

File: examples/log.rs [hunk 5]
```diff
@@ -1,99 +1,114 @@
...
+
+        // Test each log function
+        event_tests::test_log0();
+        event_tests::test_log1(value1);
+        event_tests::test_log2(value1, value2);
+        event_tests::test_log3(value1, value2, value3);
+        event_tests::test_log4(value1, value2, value3, value4);
+
+        // Test multiple logs
+        event_tests::test_multiple_logs(value1, value2, value3, value4).unwrap();
```
Comment (clearloop) [inline examples/log.rs:109]: should compile the contract and then call these interfaces to complete the tests

File: zink/src/primitives/address.rs [hunk 6]
```diff
@@ -1,19 +1,25 @@
...
 
 impl Address {
     /// Returns empty address
+    #[cfg(not(target_family = "wasm"))]
+    pub const fn empty() -> Self {
+        Address([0; 20])
+    }
+
+    /// Returns empty address
+    #[cfg(target_family = "wasm")]
```
Comment (clearloop) [inline zink/src/primitives/address.rs:20]:
```suggestion
```

File: zink/src/primitives/address.rs [hunk 7]
```diff
@@ -1,19 +1,25 @@
...
     pub const fn empty() -> Self {
-        Address(Bytes20::empty())
+        Address(0)
```
Comment (clearloop) [inline zink/src/primitives/address.rs:22]:
```suggestion
        Address(Bytes20::empty())
```

File: zink/src/primitives/address.rs [hunk 8]
```diff
@@ -28,42 +34,33 @@ impl Address {
...
+        output[12..].copy_from_slice(&self.0);
+        // Bytes32::from(U256::from(output))
+        Bytes32::empty()
     }
 }
 
 impl Asm for Address {
     fn push(self) {
-        unsafe { ffi::bytes::push_bytes20(self.0) }
+        unsafe { ffi::asm::push_address(self) }
```
Comment (clearloop) [inline zink/src/primitives/address.rs:51]:
```suggestion
        unsafe { ffi::bytes::push_bytes20(self.0) }
```

File: zink/src/primitives/address.rs [hunk 9]
```diff
@@ -28,42 +34,33 @@ impl Address {
...
     }
 
     #[cfg(not(target_family = "wasm"))]
     fn bytes32(&self) -> [u8; 32] {
-        self.0.bytes32()
+        let mut output = [0; 32];
+        output[12..].copy_from_slice(&self.0);
+        output
```
Comment (clearloop) [inline zink/src/primitives/address.rs:58]:
```suggestion
        self.0.bytes32()
```

File: zink/src/primitives/address.rs [hunk 10]
```diff
@@ -28,42 +34,33 @@ impl Address {
...
-impl From<[u8; 20]> for Address {
-    fn from(value: [u8; 20]) -> Self {
-        Address(Bytes20(value))
-    }
-}
-
-impl TransientStorageValue for Address {
-    fn tload() -> Self {
-        Address(unsafe { ffi::bytes::tload_bytes20() })
+        unsafe { ffi::asm::sload_address() }
```
Comment (clearloop) [inline zink/src/primitives/address.rs:64]:
```suggestion
        Self(unsafe { ffi::bytes::sload_bytes20() })
```

**Review by clearloop — CHANGES_REQUESTED** [review]
Your new interfaces of the example is correct! however on the other hand

1. you have modified the correct code to incorrect, the Address / Bytes conversion are correct in main
2. the tests of the logs are incorrect, you need to assert the output logs from EVM
3. once you can run log tests, you will reveal new error which requires some modifications in codegen, focus on `2.` now

File: examples/log.rs (hunk 5 above)
Comment (malik672) [reply examples/log.rs:109]: @clearloop how do we assert the output, logs don't return yet

Comment (clearloop) [reply examples/log.rs:109]: for example here:

https://github.com/zink-lang/zink/blob/main/examples/log.rs#L44-L50

1. we deploy the contract to our testing EVM
2. we call the method, in the returned `info`, there is a log field, see https://docs.zink-lang.org/rustdocs/zint/struct.Info.html#structfield.logs

**Pull request closed without merging**